doc: Add sysbuild migration guide (#15845)
nrfconnect/sdk-nrf: doc/nrf/config_and_build/sysbuild_migration.rst

+---------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_MCUBOOT_APP_SYNC_UPDATEABLE_IMAGES`            | Will set the main application number of updateable images as well as MCUboot if enabled, otherwise will only set MCUboot |
+---------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_SECURE_BOOT_MCUBOOT_VERSION`                   | MCUboot version string to use when creating MCUboot update package for application secure boot mode                      |
+---------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------+

> Support for unsigned images has been added, as well as image encryption, these options will generate the respective output files for the main application build. Any MCUboot configuration that was previously done in the main application or MCUboot needs to be updated to apply it at a sysbuild level, if this is not done then settings of these builds will be forcefully replaced with the default which sysbuild generates, which will make firmware updates incompatible with firmware images built in previous versions of |NCS|.

Reviewer: ```suggestion
Support for unsigned images and image encryption has been added.
These options generate the respective output files for the main application build.
Any MCUboot configuration that was previously done in the main application or MCUboot needs to be updated to apply at the Sysbuild level.
If this is not done, the settings of these builds will be forcefully replaced with the default generated by Sysbuild, making firmware updates incompatible with firmware images built in previous versions of |NCS|.
```

.. _child_parent_to_sysbuild_migration_secure_boot:

Secure boot
###########

> Secure boot mode selection and key file configuration is handled by sysbuild, the following Kconfig options exist:

Reviewer: Sysbuild handles the mode selection of secure boot and the configuration of the key file .
The following Kconfig options are available:

+------------------------------------------------------------+-----------------------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_SECURE_BOOT_SIGNING_COMMAND`    | Command called for custom signing, will have file to sign provided as an argument       |
+------------------------------------------------------------+-----------------------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_SECURE_BOOT_SIGNING_PUBLIC_KEY` | Absolute path to signing key public file                                                |
+------------------------------------------------------------+-----------------------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_SECURE_BOOT_PUBLIC_KEY_FILES`   | Comma-separated value list of absolute paths to signing public key files                |
+------------------------------------------------------------+-----------------------------------------------------------------------------------------+

> Secure boot now allows for being enabled centrally from sysbuild for both application and network cores for nRF53-based boards. Configuration that was done in images themselves before must now be applied at a sysbuild level, if not then the secure boot image(s) will not be built or settings of these builds will be forcefully replaced with the default which sysbuild generates, which will make firmware updates incompatible with firmware images built in previous versions of |NCS|.

Reviewer: ```suggestion
Secure boot can now be enabled centrally from Sysbuild for both the application and network cores for nRF53-based boards. 
Configuration that was previously done in the images themselves must now be applied at the Sysbuild level.
If not, the secure boot images are not built, or the settings of these builds are forcefully replaced with the default generated by Sysbuild, making firmware updates incompatible with firmware images built in previous versions of |NCS|.
```

.. _child_parent_to_sysbuild_migration_bluetooth_fast_pair:

Bluetooth fast pair
###################

> Bluetooth fast pair hex generation is now handled by sysbuild, the following Kconfig options exist:

Reviewer: ```suggestion
Sysbuild now handles Bluetooth Fast Pair HEX generation.
The following Kconfig options are available:
```

+------------------------------------------+-------------------------------------------+
| Kconfig option                           | Description                               |
+==========================================+===========================================+
| :kconfig:option:`SB_CONFIG_BT_FAST_PAIR` | Enables Bluetooth fast pair functionality |
+------------------------------------------+-------------------------------------------+

> Applications must set this at a sysbuild level if they want the Bluetooth fast pair data generated, the method of supply the key data via command line remains unchanged as it worked in previous |NCS| versions.

Reviewer: ```suggestion
Applications must set this at the Sysbuild level if you want the Bluetooth Fast Pair data generated.
The method of supplying the key data via the command line remains unchanged from previous |NCS| versions.
```

Reviewer: I think we should link to the docs explaining how to supply the key data via CLI

Author: @MarekPieta do you have a link?

.. _child_parent_to_sysbuild_migration_matter:

Matter (Connected home over IP)
###############################

> Matter configuration for generating factory data and over-the-air firmware update images is now controlled directly by sysbuild, the following Kconfigs exist:

Reviewer: ```suggestion
Sysbuild now directly controls Matter configuration for generating factory data and over-the-air firmware update images.
The following Kconfig options are available:
```

+---------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_MATTER_FACTORY_DATA`                     | Generate factory data                             |
+---------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_MATTER_FACTORY_DATA_MERGE_WITH_FIRMWARE` | Merge factory data with main application firmware |
+---------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_MATTER_OTA`                              | Generate over-the-air firmware update image       |
+---------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_MATTER_OTA_IMAGE_FILE_NAME`              | Filename for over-the-air firmware update image   |
+---------------------------------------------------------------------+---------------------------------------------------+

> Applications must enable these options if they generate factory data or need an over-the-air firmware update, note that the configuration data for the factory data file is still configured from the main application.

Reviewer: ```suggestion
Applications must enable these options if they generate factory data or need an over-the-air firmware update. 

.. note::
   The configuration data for the factory data file is still configured from the main application.
```

.. _child_parent_to_sysbuild_migration_nrf700x:

> nRF700x

Reviewer: Is there also some nRF54H20-specific info that should be added? @nordicjm

Author: No clue, nrf54h20 is out of my scope

#######

> nRF700x support for operating mode and storage of firmware has moved to sysbuild, the following Kconfigs exist:

Reviewer: ```suggestion
Support for nRF700x operating mode and firmware storage has moved to Sysbuild.
The following Kconfig options are available:
```

| :kconfig:option:`SB_CONFIG_WIFI_NRF700X_SYSTEM_WITH_RAW_MODES` | Use system with raw modes firmware patches and set application to this mode |
+----------------------------------------------------------------+-----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_WIFI_PATCHES_EXT_FLASH_DISABLED`    | Load firmware patches directly from ram (default)                           |
+----------------------------------------------------------------+-----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_WIFI_PATCHES_EXT_FLASH_XIP`         | Load firmware patches from external flash using XIP                         |
+----------------------------------------------------------------+-----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_WIFI_PATCHES_EXT_FLASH_STORE`       | Load firmware patches from external flash into RAM and load to radio        |
+----------------------------------------------------------------+-----------------------------------------------------------------------------+

> Applications must be updated to select the required Kconfigs that they need for their applications to work at a sysbuild level, these sysbuild Kconfig options are no longer defaulted or gated depending upon what features that the main application uses so need to be manually set properly. If these are not set than nRF700x functionality will not work.

Reviewer: ```suggestion
You must update your applications to select the required Kconfig options at the Sysbuild level for their applications to work.
These Sysbuild Kconfig options are no longer defaulted or gated depending on the features that the main application uses, so you must set these manually. If these options are not set, nRF700x functionality will not work.

```

Reviewer: The first sentence is still not clear "Applications must be updated to select the required Kconfigs that they need for their applications to work at a sysbuild level" what do you mean here?

.. _child_parent_to_sysbuild_migration_dfu_multi_image_build:

DFU multi image build
#####################

> 

Reviewer: ```suggestion

The following Kconfig options are available:

```

+-------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_APP`           | Include application update in package             |
+-------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_NET`           | Include network core image update in package      |
+-------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_MCUBOOT`       | Include MCUboot update in package                 |
+-------------------------------------------------------------------+---------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_MULTI_IMAGE_PACKAGE_WIFI_FW_PATCH` | Include nRF7000x firmware patch update in package |
+-------------------------------------------------------------------+---------------------------------------------------+

.. _child_parent_to_sysbuild_migration_dfu_zip:

DFU zip
#######

> 

Reviewer: ```suggestion

The following Kconfig options are available:

```

+-------------------------------------------------------------+----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_ZIP_NET`                     | Include network-core image update in zip archive                           |
+-------------------------------------------------------------+----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_ZIP_WIFI_FW_PATCH`           | Include nRF700x firmware patch update in zip archive                       |
+-------------------------------------------------------------+----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA` | Include Bluetooth mesh metadata in zip archive                             |
+-------------------------------------------------------------+----------------------------------------------------------------------------+

.. _child_parent_to_sysbuild_migration_partition_manager:

Partition manager
#################

> 

Reviewer: ```suggestion

The following Kconfig options are available:

```

+-----------------------------------------------------------------+----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_PM_MCUBOOT_PAD`                      | MCUboot image header padding                                               |
+-----------------------------------------------------------------+----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_PM_EXTERNAL_FLASH_MCUBOOT_SECONDARY` | Places the secondary MCUboot update partition in external flash            |
+-----------------------------------------------------------------+----------------------------------------------------------------------------+
| :kconfig:option:`SB_CONFIG_PM_OVERRIDE_EXTERNAL_DRIVER_CHECK`   | Will force override the external flash driver check                        |
+-----------------------------------------------------------------+----------------------------------------------------------------------------+

.. _child_parent_to_sysbuild_migration_filename_changes:

Filename changes
################

> Some output file names have changed from child/parent image or have changed the directory that they are created in, this is because sysbuild properly namespaces images in a project. The changes to final output files (ignoring artefacts and intermediary files) are as following:

Reviewer: ```suggestion
Some output file names have changed from child/parent image configurations or have changed the directory where they are created.
This is because Sysbuild properly namespaces images in a project.
The changes to final output files (ignoring artifacts and intermediary files) are as follows:
```

+-----------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| ``zephyr/dfu_application.zip``                      | ``dfu_application.zip``                                                                                                                                      |
+-----------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+
| ``zephyr/dfu_mcuboot.zip``                          | ``dfu_mcuboot.zip``                                                                                                                                          |
+-----------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+

Example output files
====================

> To demonstrate the expected output files when using sysbuild for an application build, the following shows and describes what the output files for the matter_weather_station are when building using the thingy53/nrf5340/cpaupp board target:

Reviewer: ```suggestion
To demonstrate the expected output files when using Sysbuild for an application build, the following sections show and describe the output files for the `matter_weather_station` when building using the `thingy53/nrf5340/cpaupp` board target:

```

Provision/container files

> =========================

Reviewer: ```suggestion
-------------------------
```

> 

Reviewer: ```suggestion

The expected output files are the following:

```

+-----------------------+-------------------------------------------------------+
| File                  | Description                                           |
+=======================+=======================================================+
| ``b0n_container.hex`` | Copy of ``b0n/zephyr/zephyr.hex``                     |
+-----------------------+-------------------------------------------------------+
| ``net_provision.hex`` | Provision data for the network core secure boot image |
+-----------------------+-------------------------------------------------------+

Image build files
-----------------

> 

Reviewer: ```suggestion
```suggestion

The expected output files are the following:

```
```

+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------+
| ``signed_by_b0_ipc_radio.hex``                      | Signed (with b0 signing key) network core IPC radio hex file                                          |
+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------+
| ``signed_by_b0_ipc_radio.bin``                      | Signed (with b0 signing key) network core IPC radio binary file                                       |
+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------+
| ``signed_by_mcuboot_and_b0_ipc_radio.hex``          | Signed (with b0 and MCUboot signing key) network core IPC radio update from application core hex file |
+-----------------------------------------------------+-------------------------------------------------------------------------------------------------------+

Combined files
--------------

> 

Reviewer: ```suggestion
```suggestion

The expected output files are the following:

```
```

| File                  | Description                                                                                                                                                           |
+=======================+=======================================================================================================================================================================+
| ``merged.hex``        | Merged application core hex file (contains merged contents of ``mcuboot/zephyr/zephyr.hex`` and ``matter_weather_station/zephyr/zephyr.signed.hex``)                  |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| ``merged_CPUNET.hex`` | Merged network core hex file (contains merged contents of ``net_provision.hex``, ``b0n_container.hex``, ``b0n/zephyr/zephyr.hex`` and ``signed_by_b0_ipc_radio.hex``) |
+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Update files
------------

> 

Reviewer: ```suggestion
```suggestion

The expected output files are the following:

```
```

| ``dfu_multi_image.bin``                             | DFU multi image file containing firmware update files and manifest                                       |
+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------+
| ``matter.ota``                                      | Matter over-the-air firmware update file                                                                 |
+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------+
| ``dfu_application.zip``                             | Zip file containing firmware update files and manifest                                                   |
+-----------------------------------------------------+----------------------------------------------------------------------------------------------------------+

.. _child_parent_to_sysbuild_migration_image_overlay_changes:

Image overlay configuration
###########################

> In child/parent image, an application could include additional configuration files in the ``child_image`` folder that would be applied to these images (see :ref:`ug_multi_image_permanent_changes`), this feature has been adapted in sysbuild, see :ref:`sysbuild_application_configuration` for an overview. Child/parent image configuration must be updated to be be usable from sysbuild, the way these files can be used differs. In child/parent image, there can be Kconfig fragments and board overlays which are all used, in sysbuild this is not supported, there can either be a Kconfig fragment overlay, or the whole application configuration directory can be changed. If an image application configuration directory is created then it must include all the required files for that image as none of the original application configuration files will be used. Sysbuild includes support for :ref:`application-file-suffixes` in applications and :ref:`sysbuild_file_suffixes` can also be used.

Reviewer: ```suggestion
In child/parent image configurations, an application could include additional configuration files in the ``child_image`` folder that would be applied to these images (see :ref:`ug_multi_image_permanent_changes`).
This feature has been adapted in Sysbuild; see :ref:`sysbuild_application_configuration` for an overview. 

You must update child/parent image configuration to use it with Sysbuild, as the way these files can be used differs:

* In child/parent image configurations, there can be Kconfig fragments and board overlays that are all used.
* In Sysbuild, this is not supported.
  You can either have a Kconfig fragment overlay or change the whole application configuration directory. 

If an image application configuration directory is created, it must include all the required files for that image, as none of the original application configuration files will be used.
Sysbuild includes support for :ref:`application-file-suffixes` in applications, and it can also use :ref:`sysbuild_file_suffixes`.
```

Example for MCUboot
===================

> The following table shows how to add custom MCUboot configuration for a project, the ``sysbuild`` folder would be created in the application's folder:

Reviewer: ```suggestion
The following table shows how to add custom MCUboot configuration for a project. 
The ``sysbuild`` folder should be created in the application's folder:
```

+--------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
| ``sysbuild/mcuboot/boards/nrf52840dk_nrf52840.overlay``      | DTS overlay for the ``nrf52840dk/nrf52840`` board target, note: used **instead** of app.overlay, not with as child/parent used to do      |
+--------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+
| ``sysbuild/mcuboot/boards/nrf9160dk_nrf9160_0_14_0.overlay`` | DTS overlay for the ``nrf9160dk@0.14.0/nrf9160`` board target, note: used **instead** of app.overlay, not with as child/parent used to do |
+--------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+

.. _child_parent_to_sysbuild_migration_scope_changes:

Scope changes
#############

> In child/parent images, because the application controlled all images, variables without a prefix would apply to the main application only, however this differs with sysbuild. Things like file suffixes, shields and snippets without an image prefix will be applied **globally** to all images, to apply them to just a single image they must be prefixed with the image name to apply them to. Without doing this, projects with multiple images (e.g. those with MCUboot) might find projects fail to build because of invalid configuration for other images.

Reviewer: ```suggestion
In child/parent images, the application controlled all images, so variables without a prefix would apply to the main application only.

In Sysbuild, elements like file suffixes, shields, and snippets without an image prefix will be applied *globally* to all images. 
To apply them to a single image, they must be prefixed with the image name.
Without doing this, projects with multiple images (for example, those with MCUboot) might fail to build due to invalid configuration for other images.
```

| ``-DFILE_SUFFIX=...``         | Applies to main application only | Applies to all images |
+-------------------------------+----------------------------------+-----------------------+
| ``-D<image>_FILE_SUFFIX=...`` | Applies to <image> only                                  |
+-------------------------------+----------------------------------+-----------------------+
| ``-DSNIPPET=...``             | Applies to main application only | Applies to all images |
+-------------------------------+----------------------------------+-----------------------+
| ``-D<image>_SNIPPET=...``     | Applies to <image> only                                  |
+-------------------------------+----------------------------------+-----------------------+
| ``-DSHIELD=...``              | Applies to main application only | Applies to all images |
+-------------------------------+----------------------------------+-----------------------+
| ``-D<image>_SHIELD=...``      | Applies to <image> only                                  |
+-------------------------------+----------------------------------------------------------+

> Configuration values that specify Kconfig fragment or overlay files (e.g. ``EXTRA_CONF_FILE`` and ``EXTRA_DTC_OVERLAY_FILE``) cannot be applied globally using child/parent image or sysbuild and function the same, without a prefix, they will be applied to the main application only, with a prefix will apply to that image only.

Reviewer: ```suggestion
Configuration values that specify Kconfig fragment or overlay files (for example, ``EXTRA_CONF_FILE`` and ``EXTRA_DTC_OVERLAY_FILE``) cannot be applied globally using either child/parent image or Sysbuild.
They function the same in both systems:

* Without a prefix, they will be applied to the main application only.
* With a prefix, they will apply to that specific image only.
```

.. _child_parent_to_sysbuild_migration_building:

Building with sysbuild
######################

> With |NCS| 2.7, building with west will use sysbuild by default in |NCS| repositories, to maintain compatibility with child/parent images, out-of-tree applications will not use sysbuild by default. Twister does not inherit this and will not use sysbuild unless tests are updated, CMake will not configure projects using sysbuild unless the invocation command is updated.

Reviewer: ```suggestion
With |NCS| 2.7, building with `west` will use Sysbuild by default in |NCS| repositories.
To maintain compatibility with child/parent images, out-of-tree applications will not use Sysbuild by default. 

Twister does not inherit this change and will not use Sysbuild unless tests are updated.
Similarly, CMake will not configure projects using Sysbuild unless the invocation command is updated.

```

.. tabs::

    .. group-tab:: west sysbuild as default

        Sysbuild can be set as the default when using west, this means that unless west is ran with ``--no-sysbuild`` then it will always use sysbuild

        .. code-block:: shell

            west config build.sysbuild True

    .. group-tab:: west (sysbuild)

        West can build a specific project using sysbuild with the following command:

>         .. zephyr-app-commands::

Reviewer: This directive seems to be causing an error in the docbuild

           :west-args: --no-sysbuild
           :compact:

    .. group-tab:: cmake (sysbuild)

        CMake can be used to configure a specific project using sysbuild image with the following command:

        .. zephyr-app-commands::
           :tool: cmake
           :app: <path_to_zephyr>/share/sysbuild
           :board: <board>
           :goals: build
           :gen-args: -DAPP_DIR=<app_path>
           :compact:

    .. group-tab:: cmake (child/parent image)

        CMake can be used to configure a specific project using child/parent image with the following command:

        .. zephyr-app-commands::
           :tool: cmake
           :app: <app_path>
           :board: <board>
           :goals: build
           :compact:

    .. group-tab:: twister (sysbuild)

        Twister test cases can build using sysbuild with the following:

        .. code-block:: yaml

            sysbuild: true

    .. group-tab:: twister (child/parent image)

        Twister test cases can build using child/parent image with the following:

        .. code-block:: yaml

            sysbuild: false

.. _child_parent_to_sysbuild_migration_incompatibilities:

Incompatibities
###############

> pre-built hex file, building for net core first

Author: yes
